recipe-server/normandy/recipes/models.py

```diff
@@ -373,6 +390,45 @@ def save(self, *args, **kwargs):
         self.implementation_hash = self.compute_implementation_hash()
         super().save(update_fields=['implementation_hash'])
 
+    def validate_arguments(self, arguments, old_arguments=None):
+        """
+        Test if `arguments` follows all action-specific rules.
+
+        If there is a previous version of the arguments, they should
+        be passed as `old_arguments`. Some validation rules depend on
+        the history of the arguments.
+
+        Raises `ValidationError` if any rules are violated.
+        """
+        if self.name == 'preference-experiment':
+            # Feature branch slugs should be unique within an experiment.
+            branch_slugs = set()
+            branch_values = set()
+            for branch in arguments['branches']:
+                if branch['slug'] in branch_slugs:
+                    raise ValidationError(self.errors['duplicate_branch_slug']
+                                          .format(slug=branch['slug']))
+                if branch['value'] in branch_values:
+                    raise ValidationError(self.errors['duplicate_branch_value']
+                                          .format(value=branch['value']))
+                branch_slugs.add(branch['slug'])
+                branch_values.add(branch['value'])
+
+            # Experiment slugs should be unique.
+            experiment_recipes = Recipe.objects.filter(latest_revision__action=self)
+            experiment_slugs = set()
+            if old_arguments is None:
+                # This is a newly created revision, so its slug should not be in the DB
+                experiment_slugs.add(arguments['slug'])
+            for recipe in experiment_recipes:
+                if recipe.arguments['slug'] in experiment_slugs:
+                    raise ValidationError(self.errors['duplicate_experiment_slug']
+                                          .format(slug=arguments['slug']))
+
+            # Once published, branches of a feature experiment cannot be modified.
+            if old_arguments is not None and arguments['branches'] != old_arguments['branches']:
```

Comment:
This doesn't check whether the experiment has been published, only that it has been saved. That may or may not be suitable for this check.

@gregglind Do we want branches to be modifiable after creating a recipe, but before it is enabled and sent out to anyone?